[VectorCombine] Generalize foldBitOpOfBitcasts to support more cast operations

rhyadav opened the pull request:
This patch generalizes the existing foldBitOpOfBitcasts optimization in the VectorCombine pass to handle
  additional cast operations beyond just bitcast.

  Fixes: [#146037](https://github.com/llvm/llvm-project/issues/146037)

  Summary

  The optimization now supports folding bitwise operations (AND/OR/XOR) with the following cast operations:
  - bitcast (original functionality)
  - trunc (truncate)
  - sext (sign extend)
  - zext (zero extend)

  The transformation pattern is:
  bitop(castop(x), castop(y)) -> castop(bitop(x, y))

  This reduces the number of cast instructions from 2 to 1, improving performance on targets where cast operations
  are expensive or where performing bitwise operations on narrower types is beneficial.
  
  Implementation Details

  - Renamed foldBitOpOfBitcasts to foldBitOpOfCastops to reflect broader functionality
  - Extended pattern matching to handle any CastInst operation
  - Added validation for each cast type's constraints (e.g., trunc requires source > dest)
  - Updated cost model to use the actual cast opcode
  - Preserves IR flags from original instructions
  - Handles multi-use scenarios appropriately

  Testing

  - Added comprehensive tests in test/Transforms/VectorCombine/bitop-of-castops.ll
  - Tests cover all supported cast types with all bitwise operations
  - Includes negative tests for unsupported patterns
  - All existing VectorCombine tests pass

rhyadav committed:
[VectorCombine] Generalize foldBitOpOfBitcasts to support more cast operations

This patch generalizes the foldBitOpOfBitcasts function (renamed to
foldBitOpOfCastops) to handle additional cast operations beyond just
bitcast. The optimization now supports:
- trunc (truncate)
- sext (sign extend)
- zext (zero extend)
- bitcast (original functionality)

The optimization transforms:
  bitop(cast(x), cast(y)) -> cast(bitop(x, y))

This reduces the number of cast instructions from 2 to 1, which can
improve performance on targets where cast operations are expensive or
where performing bitwise operations on narrower types is beneficial.

Changes:
- Renamed foldBitOpOfBitcasts to foldBitOpOfCastops
- Extended pattern matching to handle any CastInst
- Added validation for each cast type's constraints
- Updated cost model to use actual cast opcode
- Added comprehensive tests for all supported cast types

Fixes: #146037

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1020,8 +1086,7 @@ bool VectorCombine::scalarizeVPIntrinsic(Instruction &I) {
   // Determine scalar opcode
-  std::optional<unsigned> FunctionalOpcode =
-      VPI.getFunctionalOpcode();
+  std::optional<unsigned> FunctionalOpcode = VPI.getFunctionalOpcode();
```
(style) don't clang-format lines unrelated to a patch

rhyadav commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -808,46 +808,105 @@ bool VectorCombine::foldInsExtBinop(Instruction &I) {
+    if (SrcVecTy->getScalarSizeInBits() >= DstVecTy->getScalarSizeInBits())
+      return false;
+    break;
+  }
```
yeah, makes sense

rhyadav committed:
[VectorCombine] Generalize foldBitOpOfBitcasts to support more cast operations

	 1. Style fixes - reverted the irrelavant code updated by clang format
     2. Moved function documentation - Changed from internal comment to  comment above the function
     3. Removed debug statements - Removed debug statements, kept simplified cost debug message
     4. Removed cast validation - Deleted the entire switch statement  since Builder.CreateCast handles validation
     5. Improved cost calculations:
        - Added CostKind parameter to all cost function calls
        - Added Instruction* parameters where applicable (LHSCast, RHSCast, BinOp)
        - Calculated LHS and RHS cast costs separately for better accuracy
        - Hoisted cast cost calculations to avoid repetition
        - Used specific costs for multi-use adjustments

rhyadav committed:
[VectorCombine] Generalize foldBitOpOfBitcasts to support more cast operations

	1. Style fixes - reverted the irrelavant code updated by clang format

rhyadav committed:
[VectorCombine] Generalize foldBitOpOfBitcasts to support more cast operations

	1. Style fixes - reverted the irrelavant code updated by clang format

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1044,8 +1091,7 @@ bool VectorCombine::scalarizeVPIntrinsic(Instruction &I) {
 
-  LLVM_DEBUG(dbgs() << "Found a VP Intrinsic to scalarize: " << VPI
-                    << "\n");
+  LLVM_DEBUG(dbgs() << "Found a VP Intrinsic to scalarize: " << VPI << "\n");
```
(style) don't clang-format code unrelated to the patch

RKSimon commented on llvm/test/Transforms/VectorCombine/bitop-of-castops.ll:
```diff
@@ -0,0 +1,263 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -passes=vector-combine -S -mtriple=x86_64-- | FileCheck %s
```
move this file inside VectorCombine/X86 - target specific tests will fail on builds that don't include that target

rhyadav committed:
[VectorCombine] Generalize foldBitOpOfBitcasts to support more cast operations

	- Fixed CI qnd clang check style warnings
	- moved the test bitop-of-castops.ll inside llvm/test/Transforms/VectorCombine/X86

rhyadav committed:
fixed the last style issue

rhyadav commented:
@RKSimon i've addressed comments :

- moved the test to llvm/test/Transforms/VectorCombine/X86/bitop-of-castops.ll

> updated `LLVM_DEBUG(dbgs() << "Found a VP Intrinsic to scalarize: " << VPI
>                     << "\n");` 
> the ci style check `diff --git a/llvm/lib/Transforms/Vectorize/VectorCombine.cpp b/llvm/lib/Transforms/Vectorize/VectorCombine.cpp` now shows this as an issue `index 3fd9428d9a8b..ff1532336ae0 100644
--- a/llvm/lib/Transforms/Vectorize/VectorCombine.cpp
+++ b/llvm/lib/Transforms/Vectorize/VectorCombine.cpp
@@ -1090,8 +1090,7 @@ bool VectorCombine::scalarizeVPIntrinsic(Instruction &I) {
       (SplatCost * !Op0->hasOneUse()) + (SplatCost * !Op1->hasOneUse());
   InstructionCost NewCost = ScalarOpCost + SplatCost + CostToKeepSplats;
 
-  LLVM_DEBUG(dbgs() << "Found a VP Intrinsic to scalarize: " << VPI
-                    << "\n");
+  LLVM_DEBUG(dbgs() << "Found a VP Intrinsic to scalarize: " << VPI << "\n");
   LLVM_DEBUG(dbgs() << "Cost of Intrinsic: " << OldCost
                     << ", Cost of scalarizing:" << NewCost << "\n");
,`  i have gone with your recommendation over the ci check

dtcxzyw commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -862,8 +901,15 @@ bool VectorCombine::foldBitOpOfBitcasts(Instruction &I) {
+  Value *Result = Builder.CreateCast(CastOpcode, NewOp, I.getType());
+
+  // Preserve cast instruction flags
+  if (auto *NewCast = dyn_cast<CastInst>(Result)) {
```
Since we use `IRBuilder<InstSimplifyFolder>` in VectorCombine, the result may point to an existing cast instruction instead of `CastOp (NewOp)`. Therefore, it may propagate flags into an incorrect instruction. I recommend using `CastInst::Create` instead of `Builder.CreateCast`.

RKSimon replied:
@dtcxzyw We're already using the Builder::Create* + flags merge approach in a lot of other places in VectorCombine - do you recommend changing those as well? (In a separate patch to this one of course).

dtcxzyw replied:
> do you recommend changing those as well? (In a separate patch to this one of course)

Yeah. We can propagate the flags in the insert callback.

rhyadav replied:
thanks @dtcxzyw for catching this, i've made changes, tests all pass @RKSimon pls have a look, happy to take on the new patch for changes to rest of the VectorCombine, pls do create and assign.

rhyadav replied:
@dtcxzyw thanks for catching this, i've made the changes, all tests pass, @RKSimon pls have a look, happy to take on the rest of the changes in VectorCombine in a separate issue/patch.

rhyadav committed:
[VectorCombine] Generalize foldBitOpOfBitcasts to support more cast operations

	- Use CastInst::Create instead of Builder.CreateCast
	- Since VectorCombine uses `IRBuilder<InstSimplifyFolder>`, the `Builder.CreateCast()` call may return an existing cast instruction instead of creating a new one.
	- Regression tests all pass.

RKSimon commented on llvm/test/Transforms/VectorCombine/X86/bitop-of-castops.ll:
```diff
@@ -0,0 +1,263 @@
+; RUN: opt < %s -passes=vector-combine -S -mtriple=x86_64-- | FileCheck %s
+
+; Test bitwise operations with bitcast
+define <4 x i32> @and_bitcast_v4f32_to_v4i32(<4 x float> %a, <4 x float> %b) {
```
That seems to be the only outstanding - cheers

rhyadav committed:
[VectorCombine] Address review feedback: fix formatting and update tests

  - Remove trailing whitespace
  - Change bitcast tests to use integer-to-integer types instead of float-to-int
  - Keep one float-to-int test as negative test case

rhyadav commented:
@RKSimon fixed the ci format issue as there was a trailing space, also updated tests:

   - Keep one float-to-int bitcast test as a negative test case
  - Tests now demonstrate the optimization applies to integer bitcasts:
    
    pls approve

rhyadav commented:
thanks @RKSimon for reviewing this, all CI checks have passed, if you're planning to create issues to pass flags using `CastInst::Create` , happy to take on those, pls feel free to assign
